pkg/compactor/block_upload.go

```diff
@@ -140,26 +151,26 @@ func (c *MultitenantCompactor) FinishBlockUpload(w http.ResponseWriter, r *http.
 				message:    fmt.Sprintf("too many block upload validations in progress, limit is %d", maxConcurrency),
 				statusCode: http.StatusTooManyRequests,
 			}
-			writeBlockUploadError(err, op, "", logger, w)
+			writeBlockUploadError(err, "max concurrency was hit", logger, w)
 			return
 		}
 		// create validation file to signal that block validation has started
 		if err := c.uploadValidation(ctx, blockID, userBkt); err != nil {
-			writeBlockUploadError(err, op, "while creating validation file", logger, w)
+			writeBlockUploadError(err, "can't upload validation file", logger, w)
 			return
 		}
 		decreaseActiveValidationsInDefer = false
 		go c.validateAndCompleteBlockUpload(logger, userBkt, blockID, m, func(ctx context.Context) error {
 			defer c.blockUploadValidations.Dec()
 			return c.validateBlock(ctx, logger, blockID, m, userBkt, tenantID)
 		})
-	} else {
-		if err := c.markBlockComplete(ctx, logger, userBkt, blockID, m); err != nil {
-			writeBlockUploadError(err, op, "uploading meta file", logger, w)
-			return
-		}
-		level.Debug(logger).Log("msg", "successfully completed block upload")
 	}
+	if err := c.markBlockComplete(ctx, logger, userBkt, blockID, m); err != nil {
+		writeBlockUploadError(err, "can't mark block as complete", logger, w)
+		return
+	}
+
+	level.Info(logger).Log("msg", "successfully finished block upload", "validation_enabled", validationEnabled)
 
 	w.WriteHeader(http.StatusOK)
```

Comment:
Since `else` was removed here, we're now missing `w.WriteHeader(http.StatusOK); return` in the `validationEnabled` case, after starting goroutine. We can't mark block upload as complete just yet.

Reply:
You're right, this change was wrong. Thank you for catching. Should be fixed in https://github.com/grafana/mimir/pull/5711/commits/672b6855640cc51ece78f55ab0d67ac5dfe0070e

Reply:
Thank you!